modules/core/test/test_opencl.cpp

```diff
@@ -188,4 +188,133 @@ TEST(OCL_OpenCLExecutionContext, ScopeTest)
+            }
+
+            ocl::Device device1 = context1.device(0);
+
+            OpenCLExecutionContext ctx1 = OpenCLExecutionContext::create(context1, device1);
+            ASSERT_FALSE(ctx1.empty());
+            ctx1.bind();
+
+            p1 = context1.ptr();
+        }
+
+        {
+            ocl::Context context2 = ocl::Context::create(":GPU:1");
+            if (context2.empty())
+            {
+                context2 = ocl::Context::create(":CPU:");
+                if (context2.empty())
+                    throw SkipTestException("OpenCL GPU1/CPU devices are not available");
+            }
+
+            p2 = context2.ptr();
+        }
+
+        {
+            ocl::Context context3 = ocl::Context::create(":GPU:1");
+            if (context3.empty())
+            {
+                context3 = ocl::Context::create(":CPU:");
+                if (context3.empty())
+                    throw SkipTestException("OpenCL GPU1/CPU devices are not available");
+            }
+
+            p3 = context3.ptr();
+        }
+
+        initial_ctx.bind();
+
+        ASSERT_TRUE((reinterpret_cast<uintptr_t>(p1) == reinterpret_cast<uintptr_t>(p2)) &&
+                    (reinterpret_cast<uintptr_t>(p2) == reinterpret_cast<uintptr_t>(p3)));
+    }
+    catch (...)
+    {
+        initial_ctx.bind();  // restore
+        throw;
+    }
+}
+
+TEST(OCL_OpenCLExecutionContext, destroyUnusedExecContext)
+{
+    bool useOCL = cv::ocl::useOpenCL();
+
+    OpenCLExecutionContext initial_ctx = OpenCLExecutionContext::getCurrent();
+
+    if (!useOCL)
+    {
+        ASSERT_TRUE(initial_ctx.empty());  // Other tests should not broke global state
+        throw SkipTestException("OpenCL is not available / disabled");
+    }
+
+    ASSERT_FALSE(initial_ctx.empty());
+
+    try
+    {
+        void *p1, *p2;
+        {
+            ocl::Context context1 = ocl::Context::create(":GPU:1");
+            if (context1.empty())
+            {
+                context1 = ocl::Context::create(":CPU:");
+                if (context1.empty())
+                    throw SkipTestException("OpenCL GPU1/CPU devices are not available");
+            }
+
+            ocl::Device device1 = context1.device(0);
+
+            OpenCLExecutionContext ctx1 = OpenCLExecutionContext::create(context1, device1);
+            ASSERT_FALSE(ctx1.empty());
+            ctx1.bind();
+
+            p1 = context1.ptr();
+        }
+
+        initial_ctx.bind();
+
+        {
+            ocl::Context context2 = ocl::Context::create(":GPU:1");
+            if (context2.empty())
+            {
+                context2 = ocl::Context::create(":CPU:");
+                if (context2.empty())
+                    throw SkipTestException("OpenCL GPU1/CPU devices are not available");
+            }
+
+            p2 = context2.ptr();
+        }
+
+        ASSERT_NE(reinterpret_cast<uintptr_t>(p1), reinterpret_cast<uintptr_t>(p2));
```

Comment:
This check may fail (with both `:GPU:0` or `:GPU:1` in case of multiple OpenCL devices).

The same handle value can be reused by OpenCL runtime. There is not guarantee which forces of new handle.
It is similar to `malloc()/free()/malloc()` chain where both `malloc()` calls returns the same pointer.

So both `ASSERT_NE` and `ASSERT_EQ` checks are not valid.

---

> destroyUnusedExecContext

BTW, during implementation there was an idea to keep all created contexts till the program termination.

Reply:
Yes. This is a problem. Using a memory address to check for "sameness" is unsafe. In my testing, I saw the runtime happily allocate new `Context::Impl` in the same memory location as the prematurely destroyed one. Efficient. 😆

My challenge is that users of the SDK have almost no visibility into the `Context::Impl`. It is a private class and the public can't get to `contextId` within it. I believe that `contextId` is the best identifier with which to check sameness.

Does this team have an approach you prefer when functionality is needed for test cases but less value for the general public?

I have suggestions, in order of my preference

1. Create a public API `Context::Impl::getId()` which returns an `int`. I think this is the simplest.
2. Make a typedef contextId_t. Do `getId()` and return it. Update other code in ocl.* files to cascade the use of this typedef. However, I don't see a reasonable benefit of this approach because the codebase is using `CV_XADD()` to create/increment these context id's so there is already friction to use something other than a int-like type.
3. Create a custom `Context::operator==()`. However, this will cascade substantially more work with little benefit other than test cases.

Reply:
I would suggest to disable / remove this test for now.
To unblock fix merging.

Reply:
I believe I have to remove both new test cases. Two reasons:

1. I can't reliably use the `ptr()` to check for same or not-same. There is no mechanism for any test case to verify refcounting or sameness on `OpenCLExecutionContext`.
2. There is an issue (or side-affect) that is exposed by the already existing test cases; discovered today with deeper tracing in a debugger...

The `":GPU:1"` config in the already existing test cases is *definitely* using the 2nd GPU device in the system. Another PR needs to change the code to match your desire to run tests on systems that have only 1 GPU. I wanted to pre-emptively verify my test cases so I changed them to use the 0 device and deeply traced in the debugger. I found a side-affect. The existing test cases are affecting the state of the overall OpenCV system. This is undesired and readily invalidates other test cases (like mine).

When the existing test cases run, they call `executeUMatCall()` and it creates two UMat objects. These UMat objects increment the refcount on the `ocl::Context` that was `bind()` at the time of their creation. Unfortunately, they do *not* decrement that refcount when the two UMat go out of scope. The UMat also do *not* decrement when another `OpenCLExecutionContext.bind()`.

This UMat behavior leads to earlier test cases changing the state of both individual Context's and the global Context collection -- and these changes persist across test cases. For example, when the test `OCL_OpenCLExecutionContext, createGPU` runs, the default Context `""`  already has 5 refcounts on it that were created by the earlier test cases. This will also affect other Context that were created like `":GPU:0"` or `":GPU:1"`. The refcounts made by UMat's will persist across test cases.

This means that if another PR creates a `Context::getId()`, then all test cases that need to check for Context sameness need to be *before* all tests cases that will persist Context refcounts across test cases.

I'm starting to understand the general OpenCL approach. However, I lack similar knowledge with UMat and therefore I am unclear if the above UMat refcount side-affects are intentional or a issue themselves.

Reply:
Unfortunately UMat doesn't support context sharing/migration at all.
It is assumed that UMat is created, used and destroyed with the same active OpenCL context. UMat requires redesign to support multiple OpenCL contexts.

---

> The UMat also do not decrement when another OpenCLExecutionContext.bind()

I will take a look on this week.

---

> keep all created contexts till the program termination

One more reason for that is handling of such non-binded UMat buffers.

Reply:
OK. I'll open to separate issues to cover the topics outside this direct PR. I've also updated this PR to only have the single line change `addref()` and remove the two test cases.